applications/usb_uart_bridge/src/main.c

```diff
@@ -0,0 +1,260 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <uart.h>
+#include <nrfx.h>
+#include <string.h>
+
+/* Overriding weak function to set iSerial runtime. */
+u8_t *usb_update_sn_string_descriptor(void)
+{
+	static u8_t buf[] = "PCA20035_12PLACEHLDRS";
+
+	snprintk(&buf[9], 13, "%04X%08X",
+		(uint32_t)(NRF_FICR->DEVICEADDR[1] & 0x0000FFFF)|0x0000C000,
+		(uint32_t)NRF_FICR->DEVICEADDR[0]);
+
+	return (u8_t *)&buf;
+}
+
+#define THREAD_STACKSIZE		1024
+#define UART_BUF_SIZE			58
+#define THREAD_PRIORITY			7
+
+static K_FIFO_DEFINE(usb_0_tx_fifo);
+static K_FIFO_DEFINE(usb_1_tx_fifo);
+static K_FIFO_DEFINE(uart_0_tx_fifo);
+static K_FIFO_DEFINE(uart_1_tx_fifo);
+
+struct uart_data {
+	void *fifo_reserved;
+	u8_t buffer[UART_BUF_SIZE];
+	u16_t len;
+};
+
+static struct serial_dev {
+	struct device *dev;
+	void *peer;
+	struct k_fifo *fifo;
+	struct k_sem sem;
+	struct uart_data *rx;
+} devs[4];
+
+static void uart_interrupt_handler(void *user_data)
+{
+	struct serial_dev *dev_data = user_data;
+	struct device *dev = dev_data->dev;
+	struct serial_dev *peer_dev_data = (struct serial_dev *)dev_data->peer;
+	struct uart_data **rx = &dev_data->rx;
+
+	uart_irq_update(dev);
+
+	while (uart_irq_rx_ready(dev)) {
+		int data_length;
+
+		if (!(*rx)) {
+			(*rx) = k_malloc(sizeof(*(*rx)));
+			if ((*rx)) {
+				(*rx)->len = 0;
+			} else {
+				char dummy;
+
+				printk("Not able to allocate UART buffer, ");
+				printk("dropping oldest entry\n");
+
+				/* Drop the oldest entry. */
+				(void)k_fifo_get(dev_data->fifo, K_NO_WAIT);
+
+				/* Drop one byte to avoid spinning in an
+				 * eternal loop.
+				 */
+				uart_fifo_read(dev, &dummy, 1);
+				return;
+			}
+		}
+
+		data_length = uart_fifo_read(dev, &(*rx)->buffer[(*rx)->len],
+					     UART_BUF_SIZE - (*rx)->len);
+		(*rx)->len += data_length;
+
+		if ((*rx)->len > 0) {
+			if (((*rx)->len == UART_BUF_SIZE) ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\n') ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\r') ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\0')) {
+				k_fifo_put(peer_dev_data->fifo, (*rx));
+				k_sem_give(&peer_dev_data->sem);
+
+				(*rx) = NULL;
+			}
+		}
+	}
+
+	if (uart_irq_tx_ready(dev)) {
+		struct uart_data *buf = k_fifo_get(dev_data->fifo, K_NO_WAIT);
+		u16_t written = 0;
+
+		/* Nothing in the FIFO, nothing to send */
+		if (!buf) {
+			uart_irq_tx_disable(dev);
+			return;
+		}
+
+		while (buf->len > written) {
+			written += uart_fifo_fill(dev,
+						  &buf->buffer[written],
+						  buf->len - written);
+		}
+
+		while (!uart_irq_tx_complete(dev)) {
+			/* Wait for the last byte to get
+			 * shifted out of the module
+			 */
+		}
+
+		if (k_fifo_is_empty(dev_data->fifo)) {
+			uart_irq_tx_disable(dev);
+		}
+
+		k_free(buf);
+	}
+}
+
+void power_thread(void)
+{
+	while (1) {
+		u32_t status = NRF_POWER->USBREGSTATUS;
+
+		if (0 == (status & POWER_USBREGSTATUS_VBUSDETECT_Msk)) {
```

Comment:
There are two functions in `nrf_power` that you could use here: `nrf_power_usbregstatus_get` and `nrf_power_system_off`.

Reply:
Done